Bug 1139303 - Optimise edit mode. r=kgrandon

Cwiiis opened the pull request.

try-server-hook commented:
[<img src="https://avatars.githubusercontent.com/u/668518?v=3" alt="Cwiiis" width="50px" height="50px" /> Chris Lord (Cwiiis)](https://github.com/Cwiiis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=cf1779de1bd0)

try-server-hook commented:
[<img src="https://avatars.githubusercontent.com/u/668518?v=3" alt="Cwiiis" width="50px" height="50px" /> Chris Lord (Cwiiis)](https://github.com/Cwiiis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6b6349f856b5)

try-server-hook commented:
[<img src="https://avatars.githubusercontent.com/u/122602?v=3" alt="KevinGrandon" width="50px" height="50px" /> Kevin Grandon (KevinGrandon)](https://github.com/KevinGrandon) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=89fa6e74f862)

try-server-hook commented:
[<img src="https://avatars.githubusercontent.com/u/668518?v=3" alt="Cwiiis" width="50px" height="50px" /> Chris Lord (Cwiiis)](https://github.com/Cwiiis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=48440693fc6b)

try-server-hook commented:
[<img src="https://avatars.githubusercontent.com/u/668518?v=3" alt="Cwiiis" width="50px" height="50px" /> Chris Lord (Cwiiis)](https://github.com/Cwiiis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6f7bedc16feb)

try-server-hook commented:
[<img src="https://avatars.githubusercontent.com/u/668518?v=3" alt="Cwiiis" width="50px" height="50px" /> Chris Lord (Cwiiis)](https://github.com/Cwiiis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=c6dbaad93c8e)

try-server-hook commented:
[<img src="https://avatars.githubusercontent.com/u/668518?v=3" alt="Cwiiis" width="50px" height="50px" /> Chris Lord (Cwiiis)](https://github.com/Cwiiis) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=4381018536c9)

KevinGrandon commented on shared/elements/gaia_grid/js/grid_dragdrop.js:
```diff
@@ -439,6 +466,25 @@
       this.icon.setCoordinates(oldX, oldY);
 
+      if (!doProcessing) {
+        if (this.iconPositionTimeout === null && !this.isScrolling) {
```
Do you think there is any way that we can move this logic into its own method? Call sites would then need to call it after calling ```positionIcon```. It might make sense to break out the positioning, and rearranging of icons. Something like:

```
// Immediate processing.
this.updatePosition();
this.rearrangeGrid();

// Delayed processing.
this.updatePosition();
```

Cwiiis replied:
I agree, and this sounds like a good idea. Will see what I can do.

KevinGrandon commented on shared/elements/gaia_grid/js/items/grid_item.js:
```diff
@@ -78,6 +78,14 @@
+     * Cached values of the last transform of the element, to avoid redundant
+     * style changes.
+     */
+    lastX: -1,
```
nit: I wonder if we should use ```null``` instead? I had to stop and think for a second if we could hit the -1 case because you can drag the icon near an edge, and since you drag the center of the icon we might. It's probably fine, but might be more clear code-wise if we default to ```null```.

Cwiiis replied:
Nice spot, we should definitely do this.

Cwiiis committed:
Bug 1139303 - Optimise home-screen edit mode. r=kgrandon

This makes a number of small optimisations/changes:

- Don't set will-change on everything in edit mode.
- Cache item positions to avoid redundant style changes.
- Throttle grid rearrangement.
- More aggressively throttle successive grid rearrangements.
- (Mostly) fix jank while auto-scrolling on homescreen edit.
- Increase long-press delay in edit mode.

Collectively, these changes improve responsiveness of the homescreen
edit mode.

lightsofapollo commented:
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=5b81e3a2fb6f7d62bf0ccc3561eb8e41a69afed6
